docs/internals/optimizer.rst

```diff
@@ -65,31 +67,80 @@ A "runs" parameter of "1" will produce short but expensive code. The largest val
 Opcode-Based Optimizer Module
 =============================
 
-The opcode-based optimizer module operates on assembly. It splits the
+The opcode-based optimizer module operates on assembly code. It splits the
 sequence of instructions into basic blocks at ``JUMPs`` and ``JUMPDESTs``.
 Inside these blocks, the optimizer analyzes the instructions and records every modification to the stack,
 memory, or storage as an expression which consists of an instruction and
-a list of arguments which are pointers to other expressions. The opcode-based optimizer
-uses a component called "CommonSubexpressionEliminator" that amongst other
+a list of arguments which are pointers to other expressions.
+
+Additionally, the opcode-based optimizer
+uses a component called "CommonSubexpressionEliminator" that, amongst other
 tasks, finds expressions that are always equal (on every input) and combines
 them into an expression class. It first tries to find each new
 expression in a list of already known expressions. If no such matches are found,
 it simplifies the expression according to rules like
 ``constant + constant = sum_of_constants`` or ``X * 1 = X``. Since this is
 a recursive process, we can also apply the latter rule if the second factor
-is a more complex expression where we know that it always evaluates to one.
-Modifications to storage and memory locations have to erase knowledge about
-storage and memory locations which are not known to be different. If we first
-write to location x and then to location y and both are input variables, the
-second could overwrite the first, so we do not know what is stored at x after
-we wrote to y. If simplification of the expression ``x - y`` evaluates to a
-non-zero constant, we know that we can keep our knowledge about what is stored at ``x``.
+is a more complex expression which we know always evaluates to one.
+
+Certain optimizer steps symbolically tracks the storage and memory locations. For example, this
+information is used to compute Keccak-256 hashes that can be evaluated during compile time. Consider
+the sequence:
+
+::
+
+  PUSH 32
+  PUSH 0
+  CALLDATALOAD
+  PUSH 100
+  DUP2
+  MSTORE
+  KECCAK256
+
+or the equivalent Yul
+
+::
+
+  let x := calldataload(0)
+  mstore(x, 100)
+  let value := keccak256(x, 32)
+
+In this case, the optimizer has to 'symbolically' track the value at a memory location
+``calldataload(0)`` and then realize that the Keccak-256 hash can be evaluated at compile time. This
```

Comment:
Hmm these last 2 sentences basically repeat the sentence before the example code.